### ISSUE TITLE
TypeSystem: Support ref structs within generics

### ISSUE THREAD

**Issue by mangod9** [dotnet/runtime#65112]
This is a top-level issue to track supporting ref structs within generics. For additional details refer to https://github.com/dotnet/csharplang/issues/7608 and the older https://github.com/dotnet/csharplang/issues/1148. 

- [x] Finish support for `ref` fields - https://github.com/dotnet/runtime/issues/63768
- [x] Get design approval from all stakeholders
    - https://github.com/dotnet/runtime/pull/67129
    - https://github.com/dotnet/csharplang/pull/5689
    - https://github.com/dotnet/csharplang/pull/7625
    - Managed API https://github.com/dotnet/runtime/issues/68002
- [x] Enable CoreCLR Support
    - https://github.com/dotnet/runtime/pull/67783
    - Support may equivalence - see https://github.com/dotnet/roslyn/issues/63085
- [x] #94415
   - Support `gpAcceptByRefLike` metadata generic param attribute
   - JIT+AOT boxing elision support; generic sharing support
   - Interp boxing elision support
- [ ] Update ECMA-335 - https://github.com/dotnet/runtime/pull/103318
- [x] Add Test cases and Functional Validation
- [x] R2R Support
- [x] Allow `ref structs` to implement interfaces
   - [x] CoreCLR - Disallow `ref structs` to use DIM implementations - https://github.com/dotnet/runtime/pull/98070
   - [x] Mono - Disallow `ref structs` to use DIM implementations
- [x] Update [`System.Activator.CreateInstance<T>()`](https://learn.microsoft.com/dotnet/api/system.activator.createinstance) to include the `allows ref struct` constraint.
    - This will need to be done in all runtimes. - https://github.com/dotnet/runtime/pull/102636
- [ ] [Cecil/trimmer support for ref structs within generics](https://github.com/dotnet/runtime/issues/98519)

**Comment by mangod9** [comment]
@AaronRobinsonMSFT , assume this is ok to close or is there more work required here?

**Comment by mangod9** [comment]
@AaronRobinsonMSFT ping again. Anything remaining for this in 8?

**Comment by AaronRobinsonMSFT** [comment]
Only the bare minimum has been done. Moving this to .NET 9.

**Comment by tkreindler** [comment]
This feature request would be extremely valuable for unit testing. Right now, when using mocking frameworks like [NSubstitute](https://github.com/nsubstitute/NSubstitute) it's impossible to mock out a function that takes a ref-struct type like ReadOnlySpan as an argument.

For example doing the following gives a CS0306 error:

    interface IMessage
    {
        string DoSomething(ReadOnlySpan<byte>);
    }
    
    // Test Code
    IMessage message = Substitute.For<IMessage>();

    message.DoSomething(Arg.Any<ReadOnlySpan<byte>>()).Returns("success");

[NSubstitute](https://github.com/nsubstitute/NSubstitute/issues/656) and its parent library [Castle.Core](https://github.com/castleproject/Core/issues/453) have investigated this issue in the past and any fixes seem to have stalled out due to the difficulty of doing this without this language feature.

I believe this feature would fix this unit-testing gap and be a big improvement to the dotnet ecosystem.

**Comment by lambdageek** [comment]
cc @steveisok

**Comment by vitek-karas** [comment]
@AaronRobinsonMSFT  - did somebody look if this works with Cecil/Trimmer? It uses different type system and it could get confused on the new pattern.

**Comment by AaronRobinsonMSFT** [comment]
> @AaronRobinsonMSFT - did somebody look if this works with Cecil/Trimmer?

Not that I am aware of. Can I get someone from your side to try this out? What do we need to do to exercise this path? We have several IL based assemblies with the feature enabled we can pass to other tools.

**Comment by vitek-karas** [comment]
@sbomer - we should be able to add tests for illink - the infra has the setup for IL-based tests.
Also round-trip tests of Cecil would be helpful (or through illink, it's kind of the same thing) - to make sure trimmer doesn't corrupt the metadata around this.

**Comment by jbevain** [comment]
Cecil work is at https://github.com/jbevain/cecil/pull/945

**Comment by steveisok** [comment]
It looks like everything is done and this issue can be closed. If I'm mistaken, feel free to reopen.

**Comment by AaronRobinsonMSFT** [comment]
No, sadly this isn't done. After updating the design and ECMA, more work is needed. I moved to 10.

**Comment by mangod9** [comment]
Closing this tracking issue since most of the work is complete in 10. There is one work item still pending: https://github.com/dotnet/runtime/issues/105435, but that will be done in 11.
